### PR TITLE
Scheduler V2.0

### PR DESCRIPTION
**Pull request by rcunrau**
This version of the scheduler is much simpler than the previous version because the cron spec is parsed on the ground and distilled into <next, period, last>, i.e. the <next> time the task runs, the repeat <period> for subsequent schedules, and the <last> or expiry time of the schedule. All times are in Unix UTC timestamps. The other major change is that the scheduler task is only started once, and sleeps on a queue waiting for the scheduler service to give it work.

### CODE REVIEW

File: ex2_services/Services/source/scheduler/scheduler.c [hunk 1]
```diff
@@ -7,685 +19,169 @@
...
+            if (xQueueSendToBack(SchedulerNotificationQueue, &ctx, 0) != pdTRUE)
+                sys_log(NOTICE, "xQueueSend failed");
+        }
+    } break;
+
+    case GET_SCHEDULE: {
+        sys_log(DEBUG, "Get Schedule request received");
+        // Open schedule file. Note that it's OK if the file doesn't exist.
+        if (xSemaphoreTake(SchedLock, (TickType_t) SCHED_SEM_WAIT) == pdTRUE) {
+            int32_t fd = red_open(ScheduleFile, RED_O_RDONLY);
```
Comment (Taylorr82) [inline ex2_services/Services/source/scheduler/scheduler.c:122]: I forget if reliance edge will create the file if it doesn't exist here. I think you need to pass RED_O_CREAT

File: ex2_services/Services/source/scheduler/scheduler.c [hunk 2]
```diff
@@ -7,685 +19,169 @@
...
+            if (fd >= 0) {
+                // get file size through file stats
+                REDSTAT fstat;
+                int err = red_fstat(fd, &fstat);
+                if (err < 0) {
+                    sys_log(NOTICE, "red_fstat error: %d", red_errno);
+                    packet->data[OUT_DATA_BYTE] = red_errno;
+                    rc = SCHED_ERR_IO;
+                    packet->length = 3;
```
Comment (Taylorr82) [inline ex2_services/Services/source/scheduler/scheduler.c:131]: Should you close the file here?

**Review by Taylorr82 — COMMENTED** [review]
It's so simple now! 

I would check over the reliance edge failure paths one more time, just since I pointed out that one thing about creating the file, I don't quite see where it goes on open failure there

File: ex2_services/Services/source/scheduler/scheduler.c (hunk 1 above)
Comment (rcunrau) [reply ex2_services/Services/source/scheduler/scheduler.c:122]: This is in GET_SCHEDULE, so I don't what to create the file if it doesn't exist (see comment line 120).

File: ex2_services/Services/source/scheduler/scheduler.c (hunk 2 above)
Comment (rcunrau) [reply ex2_services/Services/source/scheduler/scheduler.c:131]: There is a "red_close(fd)" on line 156

Comment (Taylorr82) [reply ex2_services/Services/source/scheduler/scheduler.c:131]: GitHub doesn't seem to handle these large diffs very nicely, I'll pull down the changes and review them locally instead

**Pull request merged**